store/tikv/gcworker/gc_worker.go

```diff
@@ -320,6 +321,42 @@ func (w *GCWorker) checkPrepare(ctx context.Context) (bool, uint64, error) {
 	return true, oracle.ComposeTS(oracle.GetPhysical(*newSafePoint), 0), nil
 }
 
+// calculateNewSafePoint uses the current global transaction min start timestamp to calculate the new safe point.
+func (w *GCWorker) calSafePointByMinStartTS(safePoint time.Time) time.Time {
+	kvs, err := w.store.GetSafePointKV().GetWithPrefix(domain.ServerMinStartTSPath)
+	if err != nil {
+		logutil.BgLogger().Warn("get all minStartTS failed", zap.Error(err))
+		return safePoint
+	}
+	var globalMinStartTS uint64 = math.MaxUint64
+	for _, v := range kvs {
+		minStartTS, err := strconv.ParseUint(string(v.Value), 10, 64)
+		if err != nil {
+			logutil.BgLogger().Warn("parse minStartTS failed", zap.Error(err))
+			continue
+		}
+		if minStartTS != 0 && minStartTS < globalMinStartTS {
+			globalMinStartTS = minStartTS
+		}
+	}
+	physical := oracle.ExtractPhysical(globalMinStartTS)
+	sec, nsec := physical/1e3, (physical%1e3)*1e6
+	globalMinStartTime := time.Unix(sec, nsec)
+
+	diff := safePoint.Sub(globalMinStartTime)
+	maxTxnTimeUse := time.Duration(tikv.MaxTxnTimeUse)*time.Millisecond + 10*time.Second
+	logutil.BgLogger().Debug("calSafePointByMinStartTS",
+		zap.Time("safePoint", safePoint),
+		zap.Time("globalMinStartTime", globalMinStartTime),
+		zap.Duration("diff", diff))
+	// If the transaction is not too old, we could use it as the new safe point.
+	if diff < maxTxnTimeUse && globalMinStartTime.Before(safePoint) {
+		return globalMinStartTime
+	}
+
+	return safePoint
```

Comment:
It's is possible that the `diff > maxTxnTimeUse` but there is another transaction startTS before the safepoint.
So we need to check `maxTxnTimeUse` during `kvs` iteration.

Reply:
I think we needn't, since for these transactions which run longer than `maxTxnTimeUse` should be failed finally